### PR TITLE
feat: token-based authentication

### PR DESCRIPTION
**Pull request by GeorgeMac**
Supports #779 

This is a (relatively) long-standing branch and PR. It is now ready for final review.
Each of the constituent PRs were reviewed into this branch though.

Here is the overview:

1. The ability to enable authentication requirement for API routes via configuration.
```yaml
authentication:
  required: true
  methods:
    token:
      enabled: true
```
2. Initial support for static API tokens via `Authentication: Bearer ...` header.
3. When no token has been created, an initial token is created and emitted to logs (bootstrapping).

## PRs

- #1093
- #1095
- #1102
- https://github.com/flipt-io/flipt/pull/1110
- https://github.com/flipt-io/flipt/pull/1114

### CODE REVIEW

File: internal/server/auth/middleware.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+		}
+
+		auth, err := authenticator.GetAuthenticationByClientToken(ctx, clientToken)
+		if err != nil {
+			logger.Error("unauthenticated",
+				zap.String("reason", "error retrieving authentication for client token"),
+				zap.Error(err))
+			return ctx, errUnauthenticated
+		}
+
```
Comment (markphelps) [inline internal/server/auth/middleware.go:70]: should we check if the token is expired here?

File: internal/storage/auth/bootstrap.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+// Bootstrap creates an initial static authentication of type token
+// if one does not already exist.
+func Bootstrap(ctx context.Context, store storage.AuthenticationStore) (string, error) {
+	req := storage.NewListRequest(ListWithMethod(rpcauth.Method_METHOD_TOKEN))
+	set, err := store.ListAuthentications(ctx, req)
+	if err != nil {
+		return "", fmt.Errorf("bootstrapping authentication store: %w", err)
+	}
+
+	// if at-least auth token exists we do not create one
```
Comment (markphelps) [inline internal/storage/auth/bootstrap.go:20]:
```suggestion
	// if at-least 1 auth token exists we do not create one
```

File: internal/storage/auth/sql/store.go [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+
+// NewStore constructs and configures a new instance of *Store.
+// Queries are issued to the database via the provided statement builder.
+func NewStore(driver storagesql.Driver, builder sq.StatementBuilderType, logger *zap.Logger, opts ...Option) *Store {
+	store := &Store{
+		logger:  logger,
+		driver:  driver,
+		builder: builder,
+		now: func() *timestamppb.Timestamp {
+			// we truncate timestampts to the microsecond to support Postgres/MySQL
```
Comment (markphelps) [inline internal/storage/auth/sql/store.go:44]:
```suggestion
			// we truncate timestamps to the microsecond to support Postgres/MySQL
```

**Review by markphelps — COMMENTED** [review]
mostly minor nitpicks, one comment around checking expiry

File: internal/server/auth/middleware.go (hunk 1 above)
Comment (GeorgeMac) [reply internal/server/auth/middleware.go:71]: 💯 I missed this. Just added now 👍

**Review by markphelps — APPROVED** [review]
we have wift off 🚀

**Pull request merged**